File: 2024/02/haskell/02.hs
{- stack script --resolver lts-14.20 -}
module Main where

check :: [Int] -> Bool
check xs = valid $ check' 0 xs
  where
    valid :: [Int] -> Bool
    valid = all (\x -> 1 <= x && x <= 3)

    check' :: Int -> [Int] -> [Int]
    check' _ [_] = [1]
    check' dir (x:y:xs) | dir' /= 0 && dir' * dir /= -1 = del : check' dir' (y:xs)
                        | otherwise                    = [0]
      where
        dir' | x < y     = -1
             | x > y     = 1
             | otherwise = 0
        del = abs $ x - y

bool2int :: Bool -> Int
bool2int True  = 1
bool2int False = 0

task1 :: [[Int]] -> Int
task1 xs = sum $ flip map xs $ bool2int . check

--------------------------------------------------------------------------------------------------

data Zipper a = Zip {
  left  :: [a]
, focus :: a
, right :: [a]
}

zipper :: [a] -> Zipper a
zipper (x:xs) = Zip [] x xs

next :: Zipper a -> Zipper a
next (Zip ls f (r:rs)) = Zip (f:ls) r rs
next z                 = z

dampener :: [Int] -> [[Int]]
dampener xs = dampener' $ zipper xs
  where
    dampener'   (Zip ls _ []) = [reverse ls]
    dampener' z@(Zip ls f rs) = (flip (++) rs $ reverse ls) : (dampener' $ next z)

task2 xs = sum $ flip map xs $ bool2int . (\x -> if check x then True else any check $ dampener x)

main :: IO ()
main = do
  input <- fmap (fmap (fmap read . words) . lines) getContents
  putStr "Task 1: "
  print $ task1 input
  putStr "Task 2: "
  print $ task2 input

File: 2024/02/rust/02.rs
use std::io;

fn check(v: &Vec<i32>) -> bool {
    let mut os = Vec::new();
    let mut ok = true;
    for i in 1..v.len() {
        let del = (v[i] - v[i-1]).abs();
        ok &= 1 <= del && del <= 3;
        let ord : i32;
        if v[i] > v[i-1] {
            ord = 1;
        }
        else if v[i] < v[i-1] {
            ord = -1;
        }
        else {
            ord = 0;
        }
        os.push(ord);
    }
    os.iter().all(|&x| x == os[0]) && ok
}

fn task1 (input: &Vec<Vec<i32>>) -> i32 {
    input.iter().filter(|x| check(x)).count() as i32
}

fn dampener(v: &Vec<i32>) -> Vec<Vec<i32>> {
    let mut res = Vec::new();
    for i in 0..v.len() {
        let mut a = Vec::new();
        for j in 0..v.len() {
            if i != j {
                a.push(v[j]);
            }
        }
        res.push(a);
    }
    res
}

fn task2 (input: &Vec<Vec<i32>>) -> i32 {
    input.iter().filter(|x| check(x) || dampener(x).iter().any(|y| check(y))).count() as i32
}

fn main () {
    let mut input: Vec<Vec<i32>> = Vec::new();
    let mut line = String::new();
    while io::stdin().read_line(&mut line).unwrap() > 0 {
        let iter = line.trim().split_whitespace();
        let a = iter.map(|x| x.parse::<i32>().unwrap()).collect::<Vec<i32>>();
        input.push(a);
        line.clear();
    }

    println!("Task 1: {:?}", task1(&input));
    println!("Task 2: {:?}", task2(&input));
}
